test/hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java

```diff
@@ -0,0 +1,165 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+
+public class TestHugePageDecisionsAtVMStartup {
+
+    // End user warnings, printing with Xlog:pagesize at warning level, should be unconditional
+    static final String warningNoTHP = "[warning][pagesize] TransparentHugePages is not supported by the operating system.";
+    static final String warningNoLP = "[warning][pagesize] UseLargePages disabled, no large pages configured and available on the system.";
+
+    static final String buildSizeString(long l) {
+        String units[] = { "G", "M", "K" };
+        long factor = 1024 * 1024 * 1024;
+        for (int i = 2; i >= 0; i--) {
+            if (l > factor) {
+                return Long.toString(l / factor) + units[i];
+            }
+            factor /= 1024;
+        }
+        return Long.toString(l) + "B";
+    }
+
+    static void testOutput(boolean useLP, boolean useTHP, OutputAnalyzer out, HugePageConfiguration configuration) {
+
+        // Note: If something goes wrong, the JVM warns but continues, so we should never see an exit value != 0
+        out.shouldHaveExitValue(0);
+
+        // Static hugepages:
+        // Let X = the default hugepage size of the system (the one in /proc/meminfo).
+        // The JVM will cycle through page sizes, starting at X, down to the smallest hugepage size.
+        //
+        // Example 1: a system with 1GB and 2MB pages, the default hugepage size is 1GB (can only be done
+        // via kernel parameter). the JVM should first attempt to use 1GB pages, failing that should try 2MB, failing
+        // that give up and disable -UseLargePages.
+        //
+        // Example 1: same system, but the default hugepage size is 2MB. The JVM should not attempt to use 1GB pages.
+        //
+        // This picture gets more complex with -XX:LargePageSizeInBytes, which overrides the default
+        // large page size; but we ignore this for now (feel free to extend the test to cover LBSiB too).
+
+        boolean haveUsableStaticHugePages = false;
+        if (configuration.supportsStaticHugePages()) {
+            long defaultLargePageSize = configuration.getStaticDefaultHugePageSize();
+            Set<HugePageConfiguration.StaticHugePageConfig> configs = configuration.getStaticHugePageConfigurations();
+            for (HugePageConfiguration.StaticHugePageConfig config: configs) {
+                if (config.pageSize <= defaultLargePageSize) {
+                    if (config.nr_hugepages > 0 || config.nr_overcommit_hugepages > 0) {
+                        haveUsableStaticHugePages = true; break;
+                    }
+                }
+            }
+        }
+
+        if (useTHP && !useLP) {
+            useLP = true; // its implicit
+        }
+
+        if (!useLP) {
+            out.shouldContain("[info][pagesize] Large page support disabled");
+        }
+
+        else if (useLP && !useTHP &&
```

Comment:
I would prefer skipping the empty lines here and below.
```suggestion
        } else if (useLP && !useTHP &&
```